File: latex-parse/clients/javascript/latex-tokenizer.js
/**
 * LaTeX Perfectionist v25 - JavaScript Client Library
 * High-performance LaTeX tokenization via REST API
 */

class LaTeXTokenizer {
    /**
     * Create a LaTeX tokenizer client
     * @param {string} host - REST API server hostname
     * @param {number} port - REST API server port
     * @param {number} timeout - Request timeout in milliseconds
     */
    constructor(host = 'localhost', port = 8080, timeout = 30000) {
        this.baseUrl = `http://${host}:${port}`;
        this.timeout = timeout;
    }

    /**
     * Tokenize LaTeX content
     * @param {string} latexContent - LaTeX source code to tokenize
     * @returns {Promise<Object>} Tokenization result
     */
    async tokenize(latexContent) {
        const url = `${this.baseUrl}/tokenize`;

        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);

            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({ latex: latexContent }),
                signal: controller.signal
            });

            clearTimeout(timeoutId);

            if (response.status === 200) {
                return await response.json();
            } else if (response.status === 400) {
                const error = await response.json();
                throw new Error(`Invalid request: ${error.error || 'Unknown error'}`);
            } else if (response.status === 413) {
                throw new Error('LaTeX content too large (max 10MB)');
            } else if (response.status === 503) {
                throw new Error('Tokenization service unavailable');
            } else {
                throw new Error(`Unexpected status code: ${response.status}`);
            }
        } catch (error) {
            if (error.name === 'AbortError') {
                throw new Error('Request timeout');
            }
            throw error;
        }
    }

    /**
     * Tokenize LaTeX content from a file (Node.js only)
     * @param {string} filepath - Path to LaTeX file
     * @returns {Promise<Object>} Tokenization result
     */
    async tokenizeFile(filepath) {
        if (typeof window !== 'undefined') {
            throw new Error('tokenizeFile is only available in Node.js');
        }

        const fs = require('fs').promises;
        const content = await fs.readFile(filepath, 'utf-8');
        return this.tokenize(content);
    }

    /**
     * Tokenize multiple LaTeX documents in parallel
     * @param {string[]} documents - Array of LaTeX documents
     * @param {number} maxConcurrent - Maximum concurrent requests
     * @returns {Promise<Object[]>} Array of tokenization results
     */
    async batchTokenize(documents, maxConcurrent = 4) {
        const results = [];
        const queue = [...documents];
        const inProgress = [];

        while (queue.length > 0 || inProgress.length > 0) {
            // Start new requests up to the limit
            while (inProgress.length < maxConcurrent && queue.length > 0) {
                const doc = queue.shift();
                const promise = this.tokenize(doc)
                    .then(result => ({ success: true, result }))
                    .catch(error => ({ success: false, error: error.message }));
                inProgress.push(promise);
            }

            // Wait for at least one to complete
            if (inProgress.length > 0) {
                const completed = await Promise.race(inProgress);
                const index = inProgress.indexOf(completed);
                inProgress.splice(index, 1);
                results.push(await completed);
            }
        }

        return results;
    }

    /**
     * Check service health
     * @returns {Promise<Object>} Health status
     */
    async healthCheck() {
        const url = `${this.baseUrl}/health`;

        try {
            const response = await fetch(url, {
                method: 'GET',
                timeout: 5000
            });

            return await response.json();
        } catch (error) {
            return {
                status: 'unhealthy',
                service: 'not responding',
                error: error.message
            };
        }
    }

    /**
     * Get Prometheus metrics
     * @returns {Promise<string>} Prometheus metrics text
     */
    async getMetrics() {
        const url = `${this.baseUrl}/metrics`;

        try {
            const response = await fetch(url, {
                method: 'GET',
                timeout: 5000
            });

            return await response.text();
        } catch (error) {
            return '';
        }
    }

    /**
     * Benchmark tokenization performance
     * @param {string} latexContent - LaTeX content to benchmark
     * @param {number} iterations - Number of iterations
     * @returns {Promise<Object>} Performance statistics
     */
    async benchmark(latexContent, iterations = 100) {
        const latencies = [];
        let errors = 0;

        for (let i = 0; i < iterations; i++) {
            const start = performance.now();
            try {
                await this.tokenize(latexContent);
                latencies.push(performance.now() - start);
            } catch (error) {
                errors++;
            }
        }

        if (latencies.length > 0) {
            latencies.sort((a, b) => a - b);

            return {
                iterations,
                errors,
                errorRate: errors / iterations,
                latencyMs: {
                    min: latencies[0],
                    p50: latencies[Math.floor(latencies.length * 0.5)],
                    p90: latencies[Math.floor(latencies.length * 0.9)],
                    p99: latencies[Math.floor(latencies.length * 0.99)],
                    max: latencies[latencies.length - 1],
                    mean: latencies.reduce((a, b) => a + b, 0) / latencies.length
                }
            };
        } else {
            return { error: 'All requests failed' };
        }
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = LaTeXTokenizer;
}

// Example usage
if (typeof require !== 'undefined' && require.main === module) {
    (async () => {
        const tokenizer = new LaTeXTokenizer();

        // Check health
        const health = await tokenizer.healthCheck();
        console.log('Service health:', health);

        // Tokenize sample LaTeX
        const sampleLatex = `
\\documentclass{article}
\\begin{document}
\\section{Introduction}
This is a test document with $x^2 + y^2 = z^2$ math.
\\end{document}
        `;

        try {
            const result = await tokenizer.tokenize(sampleLatex);
            console.log('Tokenization result:', result);

            // Benchmark performance
            const perf = await tokenizer.benchmark(sampleLatex, 10);
            console.log(`Performance: P50=${perf.latencyMs.p50.toFixed(2)}ms`);

        } catch (error) {
            console.error('Error:', error.message);
        }
    })();
}

File: latex-parse/clients/python/latex_tokenizer.py
#!/usr/bin/env python3
"""
LaTeX Perfectionist v25 - Python Client Library
High-performance LaTeX tokenization via REST API
"""

import json
import requests
from typing import Dict, List, Optional, Union
import time


class LaTeXTokenizer:
    """Client for LaTeX Perfectionist v25 tokenization service."""

    def __init__(self, host: str = "localhost", port: int = 8080, timeout: float = 30.0):
        """
        Initialize the LaTeX tokenizer client.

        Args:
            host: REST API server hostname
            port: REST API server port
            timeout: Request timeout in seconds
        """
        self.base_url = f"http://{host}:{port}"
        self.timeout = timeout
        self.session = requests.Session()

    def tokenize(self, latex_content: str) -> Dict:
        """
        Tokenize LaTeX content using ARM NEON SIMD optimization.

        Args:
            latex_content: LaTeX source code to tokenize

        Returns:
            Dictionary containing tokens and metadata

        Raises:
            ConnectionError: If service is unavailable
            ValueError: If request is invalid
        """
        url = f"{self.base_url}/tokenize"
        payload = {"latex": latex_content}

        try:
            response = self.session.post(
                url,
                json=payload,
                timeout=self.timeout
            )

            if response.status_code == 200:
                return response.json()
            elif response.status_code == 400:
                raise ValueError(f"Invalid request: {response.json().get('error', 'Unknown error')}")
            elif response.status_code == 413:
                raise ValueError("LaTeX content too large (max 10MB)")
            elif response.status_code == 503:
                raise ConnectionError("Tokenization service unavailable")
            else:
                raise Exception(f"Unexpected status code: {response.status_code}")

        except requests.exceptions.RequestException as e:
            raise ConnectionError(f"Failed to connect to tokenizer service: {e}")

    def tokenize_file(self, filepath: str) -> Dict:
        """
        Tokenize LaTeX content from a file.

        Args:
            filepath: Path to LaTeX file

        Returns:
            Dictionary containing tokens and metadata
        """
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
        return self.tokenize(content)

    def batch_tokenize(self, documents: List[str], max_workers: int = 4) -> List[Dict]:
        """
        Tokenize multiple LaTeX documents in parallel.

        Args:
            documents: List of LaTeX documents
            max_workers: Maximum parallel requests

        Returns:
            List of tokenization results
        """
        from concurrent.futures import ThreadPoolExecutor, as_completed

        results = [None] * len(documents)

        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            future_to_index = {
                executor.submit(self.tokenize, doc): i
                for i, doc in enumerate(documents)
            }

            for future in as_completed(future_to_index):
                index = future_to_index[future]
                try:
                    results[index] = future.result()
                except Exception as e:
                    results[index] = {"error": str(e)}

        return results

    def health_check(self) -> Dict:
        """
        Check if the tokenization service is healthy.

        Returns:
            Health status dictionary
        """
        url = f"{self.base_url}/health"

        try:
            response = self.session.get(url, timeout=5.0)
            return response.json()
        except:
            return {"status": "unhealthy", "service": "not responding"}

    def get_metrics(self) -> str:
        """
        Get Prometheus metrics from the service.

        Returns:
            Prometheus metrics text
        """
        url = f"{self.base_url}/metrics"

        try:
            response = self.session.get(url, timeout=5.0)
            return response.text
        except:
            return ""

    def benchmark(self, latex_content: str, iterations: int = 100) -> Dict:
        """
        Benchmark tokenization performance.

        Args:
            latex_content: LaTeX content to benchmark
            iterations: Number of iterations

        Returns:
            Performance statistics
        """
        latencies = []
        errors = 0

        for _ in range(iterations):
            start = time.perf_counter()
            try:
                self.tokenize(latex_content)
                latencies.append((time.perf_counter() - start) * 1000)  # ms
            except:
                errors += 1

        if latencies:
            latencies.sort()
            return {
                "iterations": iterations,
                "errors": errors,
                "error_rate": errors / iterations,
                "latency_ms": {
                    "min": latencies[0],
                    "p50": latencies[len(latencies) // 2],
                    "p90": latencies[int(len(latencies) * 0.9)],
                    "p99": latencies[int(len(latencies) * 0.99)],
                    "max": latencies[-1],
                    "mean": sum(latencies) / len(latencies)
                }
            }
        else:
            return {"error": "All requests failed"}

    def close(self):
        """Close the HTTP session."""
        self.session.close()

    def __enter__(self):
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit."""
        self.close()


# Example usage
if __name__ == "__main__":
    # Create tokenizer client
    tokenizer = LaTeXTokenizer()

    # Check health
    health = tokenizer.health_check()
    print(f"Service health: {health}")

    # Tokenize sample LaTeX
    sample_latex = r"""
    \documentclass{article}
    \begin{document}
    \section{Introduction}
    This is a test document with $x^2 + y^2 = z^2$ math.
    \end{document}
    """

    try:
        result = tokenizer.tokenize(sample_latex)
        print(f"Tokenization result: {result}")

        # Benchmark performance
        perf = tokenizer.benchmark(sample_latex, iterations=10)
        print(f"Performance: P50={perf['latency_ms']['p50']:.2f}ms")

    except Exception as e:
        print(f"Error: {e}")

    tokenizer.close()